package com.andradeds.exceptions;


import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.ControllerAdvice;
import org.springframework.web.bind.annotation.ExceptionHandler;
import org.springframework.web.context.request.WebRequest;
import org.springframework.web.servlet.mvc.method.annotation.ResponseEntityExceptionHandler;

import java.util.Date;


@ControllerAdvice
public class GlobalExceptionHandler extends ResponseEntityExceptionHandler{

    @ExceptionHandler(Exception.class)
    public ResponseEntity<GenericException> globalHandlerException(Exception ex, WebRequest request) {
        GenericException internalException = new GenericException(new Date(),
                ex.getMessage(),
                request.getDescription(false),
                HttpStatus.INTERNAL_SERVER_ERROR.value());

        return new ResponseEntity<GenericException>(internalException, HttpStatus.INTERNAL_SERVER_ERROR);
    }

    @ExceptionHandler(ResourceNotFoundException.class)
    public ResponseEntity<GenericException> resourceNotFound(Exception ex, WebRequest request) {

        GenericException localException = new GenericException(new Date(),
                ex.getMessage(),
                request.getDescription(false),
                HttpStatus.BAD_REQUEST.value());

        return new ResponseEntity<>(localException, HttpStatus.NOT_FOUND);
    }

    @ExceptionHandler(UserAlreadyInUseException.class)
    public ResponseEntity<GenericException> userAlreadyInUseExeception(Exception ex, WebRequest request){

        GenericException localException = new GenericException(new Date(),
                ex.getMessage(),
                request.getDescription(false),
                HttpStatus.CONFLICT.value());

        return new ResponseEntity<>(localException, HttpStatus.CONFLICT);
    }
}
